allenact/embodiedai/mapping/mapping_utils/point_cloud_utils.py

```diff
@@ -0,0 +1,538 @@
+    """
+    # Adapted from https://github.com/devendrachaplot/Neural-SLAM.
+
+    # First compute the transformation that points undergo
+    # due to the camera's horizon
+    psi = -horizon * np.pi / 180
+    cos_psi = np.cos(psi)
+    sin_psi = np.sin(psi)
+    # fmt: off
+    horizon_transform = camera_space_xyzs.new(
+        [
+            [1, 0, 0], # unchanged
+            [0, cos_psi, sin_psi],
+            [0, -sin_psi, cos_psi,],
+        ],
+    )
+    # fmt: on
+
+    # Next compute the transformation that points undergo
+    # due to the agent's rotation about the y-axis
+    phi = -rotation * np.pi / 180
+    cos_phi = np.cos(phi)
+    sin_phi = np.sin(phi)
+    # fmt: off
+    rotation_transform = camera_space_xyzs.new(
+        [
+            [cos_phi, 0, -sin_phi],
+            [0, 1, 0], # unchanged
+            [sin_phi, 0, cos_phi],],
+    )
+    # fmt: on
+
+    # Apply the above transformations
+    view_points = (rotation_transform @ horizon_transform) @ camera_space_xyzs
+
+    # Translate the points w.r.t. the camera's position in world space.
+    world_points = view_points + camera_world_xyz[:, None]
+    return world_points
+
+
+def depth_frame_to_camera_space_xyz(
+    depth_frame: torch.Tensor, mask: Optional[torch.Tensor], fov: float = 90
+) -> torch.Tensor:
+    """Transforms a input depth map into a collection of xyz points (i.e. a
+    point cloud) in the camera's coordinate frame.
+
+    # Parameters
+    depth_frame : A square depth map, i.e. an MxM matrix with entry `depth_frame[i, j]` equaling
+        the distance from the camera to nearest surface at pixel (i,j).
+    mask : An optional boolean mask of the same size (MxM) as the input depth. Only values
+        where this mask are true will be included in the returned matrix of xyz coordinates. If
+        `None` then no pixels will be masked out (so the returned matrix of xyz points will have
+        dimension 3x(M*M)
+    fov: The field of view of the camera.
+
+    # Returns
+
+    A 3xN matrix with entry [:, i] equalling a the xyz coordinates (in the camera's coordinate
+    frame) of a point in the point cloud corresponding to the input depth frame.
+    """
+    assert (
+        len(depth_frame.shape) == 2 and depth_frame.shape[0] == depth_frame.shape[1]
+    ), f"depth has shape {depth_frame.shape}, we only support (N, N) shapes for now."
+
+    resolution = depth_frame.shape[0]
+    if mask is None:
+        mask = torch.ones_like(depth_frame, dtype=bool)
+
+    # pixel centers
+    camera_space_yx_offsets = (
+        torch.stack(torch.where(mask))
+        + 0.5  # Offset by 0.5 so that we are in the middle of the pixel
+    )
+
+    # Subtract center
+    camera_space_yx_offsets -= resolution / 2.0
+
+    # Make "up" in y be positive
+    camera_space_yx_offsets[0, :] *= -1
+
+    # Put points on the clipping plane
+    camera_space_yx_offsets *= (2.0 / resolution) * math.tan((fov / 2) / 180 * math.pi)
```

Comment:
Would it make sense to consider an anamorphic camera? We might want to backproject after e.g. squeezing an image horizontally. Otherwise, it LGTM.

Reply:
I do like the idea of having an anamorphic camera. So I don't introduce new bugs I think I'll keep as it is for now and extend it if/when we need it.